src/Fixer/Phpdoc/NoSuperfluousPhpdocTagsFixer.php

```diff
@@ -0,0 +1,264 @@
+            }
+
+            $beforeArgumentIndex = $tokens->getPrevTokenOfKind($index, ['(', ',']);
+            $typeIndex = $tokens->getNextMeaningfulToken($beforeArgumentIndex);
+
+            if ($typeIndex !== $index) {
+                $info = $this->parseTypeHint($tokens, $typeIndex);
+            } else {
+                $info = [
+                    'type' => null,
+                    'allows_null' => true,
+                ];
+            }
+
+            if (!$info['allows_null']) {
+                $nextIndex = $tokens->getNextMeaningfulToken($index);
+                if (
+                    $tokens[$nextIndex]->equals('=')
+                    && $tokens[$tokens->getNextMeaningfulToken($nextIndex)]->equals([T_STRING, 'null'])
+                ) {
+                    $info['allows_null'] = true;
+                }
+            }
+
+            $argumentsInfo[$token->getContent()] = $info;
+        }
+
+        return $argumentsInfo;
+    }
+
+    private function getReturnTypeInfo(Tokens $tokens, $closingParenthesisIndex)
+    {
+        $colonIndex = $tokens->getNextMeaningfulToken($closingParenthesisIndex);
+        if ($tokens[$colonIndex]->isGivenKind(CT::T_TYPE_COLON)) {
+            return $this->parseTypeHint($tokens, $tokens->getNextMeaningfulToken($colonIndex));
+        }
+
+        return [
+            'type' => null,
+            'allows_null' => true,
+        ];
+    }
+
+    private function parseTypeHint(Tokens $tokens, $index)
+    {
+        $allowsNull = false;
+        if ($tokens[$index]->isGivenKind(CT::T_NULLABLE_TYPE)) {
+            $allowsNull = true;
+            $index = $tokens->getNextMeaningfulToken($index);
+        }
+
+        $type = '';
+        while ($tokens[$index]->isGivenKind([T_NS_SEPARATOR, T_STRING])) {
+            $type .= $tokens[$index]->getContent();
+
+            $index = $tokens->getNextMeaningfulToken($index);
+        }
+
+        return [
+            'type' => $type,
+            'allows_null' => $allowsNull,
+        ];
+    }
+
+    private function annotationIsSuperfluous(Annotation $annotation, array $info, array $symbolShortNames)
+    {
+        if ('param' === $annotation->getTag()->getName()) {
+            $regex = '/@param\s+(?:\S|\s(?!\$))+\s\$\S+\s+\S/';
+        } else {
+            $regex = '/@return\s+\S+\s+\S/';
+        }
+
+        if (Preg::match($regex, $annotation->getContent())) {
+            return false;
+        }
+
+        $annotationTypes = $this->toComparableNames($annotation->getTypes(), $symbolShortNames);
+
+        if (['mixed'] === $annotationTypes && null === $info['type']) {
+            return true;
+        }
+
+        $actualTypes = [$info['type']];
+        if ($info['allows_null']) {
+            $actualTypes[] = 'null';
+        }
+
+        return $annotationTypes === $this->toComparableNames($actualTypes, $symbolShortNames);
+    }
+
+    private function toComparableNames(array $types, array $symbolShortNames)
```

Comment:
what are the input arrays?
what is the output ?

Reply:
PHPDoc added.